delegate-app/Main.hs

```diff
@@ -1,49 +1,133 @@
 eventsProducer thisIsFirstRun =
   if thisIsFirstRun
     then return $ Just Start
     else return Nothing
 
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
+delegateServerApp ::
+  (MonadIO m, MonadTracer DelegateServerLog m) =>
+  Int ->
+  ServerAppT m
+delegateServerApp tick pending = do
+  connection <- liftIO $ acceptRequest pending
+  trace FrontendConnected
+  -- FIXME: we will probably have to do this in parallel, but how can we enqueue in the same
+  -- state? probably have to remodel this
+  execDelegateRunnerT $ flip evalStateT True $ mkRunner tick connection
+
+-- FIXME: remove concrete monads and make this tagless final (needs MonadDelegate)
+mkRunner ::
+  (MonadIO m, MonadTracer DelegateServerLog m) =>
+  Int ->
+  Connection ->
+  StateT Bool (DelegateRunnerT m) ()
+-- FIXME: we need to abort at some point but this doesn't seem
+-- to be implemented yet so we just go on
+mkRunner tick con = forever $ do
+  isFirstRun <- get
+  _ <- lift $ do
+    let encodeSend = liftIO . sendTextData con . encode
+    liftIO (receiveData con)
+      >>= ( \case
+              Nothing -> lift $ trace (DelegateError FrontendNoParse)
+              Just req -> do
+                lift . trace $ FrontendInput req
+                step <- delegateStep (FrontendRequest req)
+                lift $ traverse_ (trace . DelegateOutput) step
+                encodeSend step
+          )
+        -- yes, this is not nice, but hlint wants it so badly
+        . decode
 
-delegateTick :: Int
-delegateTick = 1_000
+    -- FIXME: more elaborate logic for eventsProducer
+    mdelegateEvent <- eventsProducer isFirstRun
+    whenJust (DelegateEvent <$> mdelegateEvent) $
+      delegateStep >=> encodeSend
+
+  liftIO $ threadDelay tick
+  put False
+
+runDelegateServer ::
+  DelegateServerConfig ->
+  TracerT DelegateServerLog IO ()
+runDelegateServer conf = do
+  trace (Started $ dlgt'port conf)
+  let fallback :: Application
+      fallback _req res =
+        res $
+          responseLBS status200 [("Content-Type", "text/plain")] "Websocket endpoint of delegate server"
+
+  tracer <- askTracer
+  liftIO $
+    run (fromIntegral $ dlgt'port conf) $
+      flip (websocketsOr defaultConnectionOptions) fallback $
+        runWithTracer' tracer . delegateServerApp (dlgt'tick conf)
 
 main :: IO ()
 main = do
   -- FIXME: cover either case. It probably should not be transformer.
-  _ <- execDelegateRunnerT $ go True
-  return ()
-  where
-    go thisIsFirstRun = do
-      mEvent <- eventsProducer thisIsFirstRun
-      consumer (DelegateEvent <$> mEvent)
-      liftIO $ threadDelay delegateTick
-      go False
+  port <- lookupEnv "port"
+
+  -- FIXME: do we need this?
+  let host :: IP = "127.0.0.1"
```

Comment:
I think while style is better cuz it makes clear that no monadic side-effects are happening.

Reply:
We can parse these as args (we will need it also for hydra node later). We can use a similar style as we use in the CLI

Reply:
> We can parse these as args (we will need it also for hydra node later). We can use a similar style as we use in the CLI

Is this in scope of this PR?

Reply:
I guess we can separate this to follow-up issue to iterate quickly.

Reply:
I opened #243 to track this